File: dexfile/src/main/java/com/github/netomi/bat/dexfile/EncodedMethod.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile;

import com.github.netomi.bat.dexfile.io.DexDataInput;
import com.github.netomi.bat.dexfile.io.DexDataOutput;
import com.github.netomi.bat.dexfile.visitor.CodeVisitor;
import com.github.netomi.bat.dexfile.visitor.DataItemVisitor;

import java.util.Objects;

/**
 * A class representing an encoded method inside a dex file.
 *
 * @see <a href="https://source.android.com/devices/tech/dalvik/dex-format#encoded-method-format">encoded method @ dex format</a>
 *
 * @author Thomas Neidhart
 */
public class EncodedMethod
extends      DexContent
{
    private int deltaMethodIndex; // uleb128
    private int methodIndex;
    private int accessFlags;      // uleb128
    private int codeOffset;       // uleb128

    public  Code code;

    public static EncodedMethod readContent(DexDataInput input, int lastIndex) {
        EncodedMethod encodedMethod = new EncodedMethod();
        encodedMethod.read(input);
        encodedMethod.updateMethodIndex(lastIndex);
        return encodedMethod;
    }

    private EncodedMethod() {
        methodIndex = DexConstants.NO_INDEX;
        accessFlags = 0;
        code        = null;
    }

    public EncodedMethod(int methodIndex) {
        this.methodIndex = methodIndex;
    }

    public int getCodeOffset() {
        return codeOffset;
    }

    public int getAccessFlags() {
        return accessFlags;
    }

    public int getMethodIndex() {
        return methodIndex;
    }

    public MethodID getMethodID(DexFile dexFile) {
        return dexFile.getMethodID(methodIndex);
    }

    public ProtoID getProtoID(DexFile dexFile) {
        return getMethodID(dexFile).getProtoID(dexFile);
    }

    public String getClassType(DexFile dexFile) {
        return dexFile.getMethodID(methodIndex).getClassType(dexFile).getType(dexFile);
    }

    public String getName(DexFile dexFile) {
        return getMethodID(dexFile).getName(dexFile);
    }

    public String getShortyType(DexFile dexFile) {
        return getMethodID(dexFile).getProtoID(dexFile).getShorty(dexFile);
    }

    public String getDescriptor(DexFile dexFile) {
        ProtoID protoID = getProtoID(dexFile);
        return protoID.getDescriptor(dexFile);
    }

    public boolean isStatic() {
        return (accessFlags & DexConstants.ACC_STATIC) != 0;
    }

    @Override
    protected void read(DexDataInput input) {
        deltaMethodIndex = input.readUleb128();
        accessFlags      = input.readUleb128();
        codeOffset       = input.readUleb128();
    }

    @Override
    protected void readLinkedDataItems(DexDataInput input) {
        if (codeOffset != 0) {
            input.setOffset(codeOffset);

            code = Code.readItem(input);
            code.readLinkedDataItems(input);
        }
    }

    private void updateMethodIndex(int lastIndex) {
        methodIndex = deltaMethodIndex + lastIndex;
    }

    private void updateDeltaMethodIndex(int lastIndex) {
        deltaMethodIndex = methodIndex - lastIndex;
    }

    @Override
    protected void updateOffsets(DataItem.Map dataItemMap) {
        codeOffset = dataItemMap.getOffset(code);
    }

    protected int write(DexDataOutput output, int lastIndex) {
        updateDeltaMethodIndex(lastIndex);
        write(output);
        return methodIndex;
    }

    @Override
    protected void write(DexDataOutput output) {
        output.writeUleb128(deltaMethodIndex);
        output.writeUleb128(accessFlags);
        output.writeUleb128(codeOffset);
    }

    public void codeAccept(DexFile dexFile, ClassDef classDef, CodeVisitor visitor) {
        if (code != null) {
            visitor.visitCode(dexFile, classDef, this, code);
        }
    }

    @Override
    public void dataItemsAccept(DexFile dexFile, DataItemVisitor visitor) {
        if (code != null) {
            visitor.visitCode(dexFile, this, code);
            code.dataItemsAccept(dexFile, visitor);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        EncodedMethod other = (EncodedMethod) o;
        return methodIndex == other.methodIndex &&
               accessFlags == other.accessFlags &&
               Objects.equals(code, other.code);
    }

    @Override
    public int hashCode() {
        return Objects.hash(methodIndex, accessFlags, code);
    }

    @Override
    public String toString() {
        return String.format("EncodedMethod[methodIndex=%d,accessFlags=%04x]", methodIndex, accessFlags);
    }
}

File: dexfile/src/main/kotlin/com/github/netomi/bat/dexfile/EncodedMethod.kt
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile

import com.github.netomi.bat.dexfile.annotation.Annotation
import com.github.netomi.bat.dexfile.annotation.visitor.AnnotationSetVisitor
import com.github.netomi.bat.dexfile.annotation.visitor.allAnnotations
import com.github.netomi.bat.dexfile.annotation.visitor.annotationCollector
import com.github.netomi.bat.dexfile.io.DexDataInput
import com.github.netomi.bat.dexfile.io.DexDataOutput
import com.github.netomi.bat.dexfile.util.DexType
import com.github.netomi.bat.dexfile.visitor.CodeVisitor
import com.github.netomi.bat.dexfile.visitor.DataItemVisitor
import com.github.netomi.bat.dexfile.visitor.PropertyAccessor
import com.github.netomi.bat.dexfile.visitor.ReferencedIDVisitor
import java.util.*

/**
 * A class representing an encoded method inside a dex file.
 *
 * @see [encoded method @ dex format](https://source.android.com/devices/tech/dalvik/dex-format.encoded-method-format)
 */
class EncodedMethod private constructor(methodIndex: Int = NO_INDEX,
                                        accessFlags: Int = 0,
                                        code       : Code = Code.empty()): EncodedMember(accessFlags) {

    private var deltaMethodIndex = 0

    var methodIndex: Int = methodIndex
        private set

    var modifiers: Set<MethodModifier> = MethodModifier.setOf(accessFlags)
        private set

    override fun updateModifiers(accessFlags: Int) {
        modifiers = MethodModifier.setOf(accessFlags)
    }

    var codeOffset = 0
        private set

    var code: Code = code
        internal set

    fun getMethodID(dexFile: DexFile): MethodID {
        return dexFile.getMethodID(methodIndex)
    }

    fun getProtoID(dexFile: DexFile): ProtoID {
        return getMethodID(dexFile).getProtoID(dexFile)
    }

    fun getClassType(dexFile: DexFile): DexType {
        return dexFile.getMethodID(methodIndex).getClassTypeID(dexFile).getType(dexFile)
    }

    override fun getName(dexFile: DexFile): String {
        return getMethodID(dexFile).getName(dexFile)
    }

    fun getShortyType(dexFile: DexFile): String {
        return getMethodID(dexFile).getProtoID(dexFile).getShorty(dexFile)
    }

    fun getParameterTypes(dexFile: DexFile): List<DexType> {
        return getProtoID(dexFile).getParameterTypes(dexFile)
    }

    fun getReturnType(dexFile: DexFile): DexType {
        return getProtoID(dexFile).getReturnType(dexFile)
    }

    fun getDescriptor(dexFile: DexFile): String {
        val protoID = getProtoID(dexFile)
        return protoID.getDescriptor(dexFile)
    }

    val isStatic: Boolean
        get() = modifiers.contains(MethodModifier.STATIC)

    val isAbstract: Boolean
        get() = modifiers.contains(MethodModifier.ABSTRACT)

    val isNative: Boolean
        get() = modifiers.contains(MethodModifier.NATIVE)

    val isPrivate: Boolean
        get() = visibility == Visibility.PRIVATE

    val isConstructor: Boolean
        get() = modifiers.contains(MethodModifier.CONSTRUCTOR)

    val isDirectMethod: Boolean
        get() = isStatic || isPrivate || isConstructor

    internal fun sort() {
        code.sort()
    }

    override fun read(input: DexDataInput) {
        deltaMethodIndex = input.readUleb128()
        accessFlags      = input.readUleb128()
        codeOffset       = input.readUleb128()
    }

    override fun readLinkedDataItems(input: DexDataInput) {
        if (codeOffset != 0) {
            input.offset = codeOffset
            code = Code.read(input)
            code.readLinkedDataItems(input)
        }
    }

    private fun updateMethodIndex(lastIndex: Int) {
        methodIndex = deltaMethodIndex + lastIndex
    }

    private fun updateDeltaMethodIndex(lastIndex: Int) {
        deltaMethodIndex = methodIndex - lastIndex
    }

    override fun updateOffsets(dataItemMap: DataItem.Map) {
        codeOffset = dataItemMap.getOffset(code)
    }

    fun write(output: DexDataOutput, lastIndex: Int): Int {
        updateDeltaMethodIndex(lastIndex)
        write(output)
        return methodIndex
    }

    override fun write(output: DexDataOutput) {
        output.writeUleb128(deltaMethodIndex)
        output.writeUleb128(accessFlags)
        output.writeUleb128(codeOffset)
    }

    fun codeAccept(dexFile: DexFile, classDef: ClassDef, visitor: CodeVisitor) {
        if (!isAbstract) {
            visitor.visitCode(dexFile, classDef, this, code)
        }
    }

    fun annotations(dexFile: DexFile, classDef: ClassDef): List<Annotation> {
        val collector = annotationCollector()
        annotationSetAccept(dexFile, classDef, allAnnotations(collector))
        return collector.items()
    }

    fun annotationSetAccept(dexFile: DexFile, classDef: ClassDef, visitor: AnnotationSetVisitor) {
        classDef.methodAnnotationSetAccept(dexFile, classDef, this, visitor)
    }

    override fun dataItemsAccept(dexFile: DexFile, visitor: DataItemVisitor) {
        visitor.visitCode(dexFile, this, code)
        code.dataItemsAccept(dexFile, visitor)
    }

    internal fun referencedIDsAccept(dexFile: DexFile, visitor: ReferencedIDVisitor)
    {
        visitor.visitMethodID(dexFile, PropertyAccessor({ methodIndex }, { methodIndex = it }))
        code.referencedIDsAccept(dexFile, visitor)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val o = other as EncodedMethod

        return methodIndex == o.methodIndex &&
               accessFlags == o.accessFlags &&
               code        == o.code
    }

    override fun hashCode(): Int {
        return Objects.hash(methodIndex, accessFlags, code)
    }

    override fun toString(): String {
        return "EncodedMethod[methodIndex=%d,accessFlags=%04x,code=%d units]".format(methodIndex, accessFlags, code.insSize)
    }

    companion object {
        fun of(methodIndex: Int, visibility: Visibility, vararg modifiers: MethodModifier): EncodedMethod {
            require(methodIndex >= 0) { "methodIndex must not be negative" }
            var accessFlags = visibility.flagValue
            for (modifier in modifiers) {
                accessFlags = accessFlags or modifier.flagValue
            }
            return EncodedMethod(methodIndex, accessFlags)
        }

        fun of(methodIndex: Int, accessFlags: Int): EncodedMethod {
            require(methodIndex >= 0) { "methodIndex must not be negative" }
            return EncodedMethod(methodIndex, accessFlags)
        }

        internal fun read(input: DexDataInput, lastIndex: Int): EncodedMethod {
            val encodedMethod = EncodedMethod()
            encodedMethod.read(input)
            encodedMethod.updateMethodIndex(lastIndex)
            return encodedMethod
        }
    }
}